### PR TITLE
Add services API

### PR DESCRIPTION
**Pull request by aborilov**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Documentation**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/apis/fake/v1/db_types.go [hunk 1]
```diff
@@ -43,6 +43,9 @@ type DBSpec struct {
 type OperationFlagType string
 
 func (o OperationFlagType) Enabled() bool {
+	if o == "" {
```
Comment (jiachengxu) [inline pkg/apis/fake/v1/db_types.go:46]: Is this needed for this PR? Otherwise, please put it in a dedicated PR

File: pkg/apiserver/internal/util/util.go [hunk 2]
```diff
@@ -28,3 +32,95 @@ func TimestampProto(t *metav1.Time) (*timestamp.Timestamp, error) {
 	}
 	return ptypes.TimestampProto(t.Time)
 }
+
+func FromMetav1(obj metav1.ObjectMeta) (*v1.ObjectMeta, error) {
```
Comment (jiachengxu) [inline pkg/apiserver/internal/util/util.go:36]: This function is not used anywhere, and also I think these helper functions should be versioned under `/v1` folder?
If this function is not used in this PR, please remove it, I have added a similar helper function in the refactor PR to share between all our services.

File: pkg/apiserver/internal/v1/service.go [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+	}
+}
+func (o serviceServer) Create(ctx context.Context, req *v1.ServiceCreateRequest) (res *v1.Service, err error) {
+	obj := &unstructured.Unstructured{}
+
+	gvk := o.gvkFromService(req.Service, req.Version)
+	obj.SetGroupVersionKind(gvk)
+	val := map[string]interface{}{}
+	if err := json.Unmarshal([]byte(req.Spec.Spec), &val); err != nil {
+		return nil, status.Errorf(codes.Internal, "creating services: spec should by type of map[string]intreface{}")
```
Comment (jiachengxu) [inline pkg/apiserver/internal/v1/service.go:60]:
```suggestion
		return nil, status.Error(codes.Internal, "creating service: spec should be type of map[string]intreface{}")
```

File: pkg/apiserver/internal/v1/service.go [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+	if err := unstructured.SetNestedMap(obj.Object, val, "spec"); err != nil {
+		return nil, status.Errorf(codes.Internal, fmt.Sprintf("creating services: %s", err.Error()))
+	}
+	if err := util.SetMetadata(obj, req.Spec.Metadata); err != nil {
+		return nil, status.Errorf(codes.Internal, fmt.Sprintf("creating services: %s", err.Error()))
+	}
+	if err := o.client.Create(ctx, obj); err != nil {
+		return nil, status.Errorf(codes.Internal, fmt.Sprintf("creating services: %s", err.Error()))
+	}
+	if err = o.client.Get(ctx, types.NamespacedName{
```
Comment (jiachengxu) [inline pkg/apiserver/internal/v1/service.go:71]: Why do you need to get the object after creating it?

File: pkg/apiserver/internal/v1/service.go [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+		return nil, status.Errorf(codes.Internal, fmt.Sprintf("converting Service: %s", err.Error()))
+	}
+	return
+}
+
+func (o serviceServer) Delete(ctx context.Context, req *v1.ServiceDeleteRequest) (*empty.Empty, error) {
+	obj := &unstructured.Unstructured{}
+	gvk := o.gvkFromService(req.Service, req.Version)
+	obj.SetGroupVersionKind(gvk)
+	if err := o.client.Get(ctx, types.NamespacedName{
```
Comment (jiachengxu) [inline pkg/apiserver/internal/v1/service.go:137]: Why do you need to `Get` before `Delete`?

File: pkg/internal/resources/elevator/elevator.go [hunk 6]
```diff
@@ -263,6 +263,30 @@ func Manifests(c Config) ([]unstructured.Unstructured, error) {
...
+			Name: c.TenantPlural + "." + c.TenantGroup,
+			Labels: map[string]string{
+				"kubecarrier.io/apiserver": "true",
+			},
+		},
+		Rules: []rbacv1.PolicyRule{
+			{
+				APIGroups: []string{c.TenantGroup},
+				Resources: []string{c.TenantPlural},
+				Verbs:     []string{"create", "delete", "get", "list", "watch"},
```
Comment (jiachengxu) [inline pkg/internal/resources/elevator/elevator.go:281]: "watch" is not needed

File: pkg/apiserver/internal/v1/service.go (hunk 4 above)
Comment (aborilov) [reply pkg/apiserver/internal/v1/service.go]: thanks, didn't know that `Create` update `obj`

File: pkg/apiserver/internal/v1/service.go (hunk 5 above)
Comment (aborilov) [reply pkg/apiserver/internal/v1/service.go]: you're right, we don't need that

File: pkg/apiserver/internal/util/util.go (hunk 2 above)
Comment (aborilov) [reply pkg/apiserver/internal/util/util.go]: done

File: pkg/apis/fake/v1/db_types.go (hunk 1 above)
Comment (aborilov) [reply pkg/apis/fake/v1/db_types.go]: will add in separate PR

File: pkg/internal/resources/elevator/elevator.go (hunk 6 above)
Comment (aborilov) [reply pkg/internal/resources/elevator/elevator.go]: done

**Comment by aborilov** [comment]
> Can you please add some unit tests and e2e tests for the new service?

of course but I thought we want this for the demo.
tests are my next step

**Comment by thetechnick** [comment]
@aborilov I would like to standardize our naming convention, because with CRDs, Services and Offerings, there are a lot of dynamic things happening.

For instances of an offering, I would like to make sure we refer to them as "Service Instances" or just Instances in the documentation and also in the API.

So could you please rename the service and endpoint to `Instance*` / `/instance`?

File: pkg/apiserver/api/v1/instance.proto [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+import "google/api/annotations.proto";
+import "google/protobuf/empty.proto";
+
+import "meta.proto";
+
+message Instance {
+  ObjectMeta metadata = 1;
+  string spec = 2;
+  string status = 3;
```
Comment (nmiculinic) [inline pkg/apiserver/api/v1/instance.proto:29]: This needs more information to be useful:
* maybe reference to the offering? it would satisfy most requirements since I could easily fetch the offering and find out the details
* instance image, provider, readiness status (boolean).

File: pkg/apiserver/api/v1/instance.proto [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+message InstanceDeleteRequest {
+  string instance = 1;
+  string version = 2;
+  string name = 3;
+  string account = 4;
+}
+
+message InstanceListRequest {
+  string instance = 1;
```
Comment (nmiculinic) [inline pkg/apiserver/api/v1/instance.proto:52]: Why are you requiring instance string in the list request; what's the purpose of it?

File: pkg/apiserver/api/v1/instance.proto [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+
+message InstanceGetRequest {
+  string instance = 1;
+  string version = 2;
+  string name = 3;
+  string account = 4;
+}
+
+message InstanceDeleteRequest {
+  string instance = 1;
```
Comment (nmiculinic) [inline pkg/apiserver/api/v1/instance.proto:45]: What's the difference between `instance` and `name`?

File: pkg/apiserver/api/v1/instance.proto [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+  string account = 3;
+  string labelSelector = 4;
+  int64 limit = 5;
+  string continue = 6;
+}
+
+message InstanceCreateRequest {
+  string instance = 1;
+  string version = 2;
+  Instance spec = 3;
```
Comment (nmiculinic) [inline pkg/apiserver/api/v1/instance.proto:63]: Isn't everything you need contained in the Instance? (except account maybe)

File: pkg/apiserver/api/v1/instance.proto [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+message InstanceGetRequest {
+  string instance = 1;
+  string version = 2;
+  string name = 3;
+  string account = 4;
+}
+
+message InstanceDeleteRequest {
+  string instance = 1;
+  string version = 2;
```
Comment (nmiculinic) [inline pkg/apiserver/api/v1/instance.proto:46]: What has the version got to do with the deletion request? It shouldn't matter since the version is useful only for get/listing, for deletion is irrelevant.

File: pkg/apiserver/api/v1/instance.proto [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+  string status = 3;
+}
+
+message InstanceList {
+  ListMeta metadata = 1;
+  repeated Instance items = 2;
+}
+
+message InstanceGetRequest {
+  string instance = 1;
```
Comment (nmiculinic) [inline pkg/apiserver/api/v1/instance.proto:38]: What's the difference between `instance` and `name`?

Comment (aborilov) [reply pkg/apiserver/api/v1/instance.proto:39]: so, `instance` it's service/offering name, i.e. `couchdbs.eu-west-1.team-a`, `name` that we get from `offering.spec.crd.name`. we can rename this field to `offering` or back to `service`.
and `name` is the name of the instance that created from this offering, like `ProdCouchDB`

File: pkg/apiserver/api/v1/instance.proto (hunk 9 above)
Comment (aborilov) [reply pkg/apiserver/api/v1/instance.proto:50]: same here
so, `instance` it's service/offering name, i.e. `couchdbs.eu-west-1.team-a`, `name` that we get from `offering.spec.crd.name`. we can rename this field to `offering` or back to `service`.
and `name` is the name of the instance that created from this offering, like `ProdCouchDB`

File: pkg/apiserver/api/v1/instance.proto (hunk 8 above)
Comment (aborilov) [reply pkg/apiserver/api/v1/instance.proto:61]: to know which instance types do ween need to list, same as `kubectl get pod`. in our case it will be `couchdbs.eu-west-1.team-a`

File: pkg/apiserver/api/v1/instance.proto (hunk 7 above)
Comment (aborilov) [reply pkg/apiserver/api/v1/instance.proto:29]: UI can get corresponding `offering` using `instance`

File: pkg/apiserver/api/v1/instance.proto (hunk 11 above)
Comment (aborilov) [reply pkg/apiserver/api/v1/instance.proto:52]: to be consistent with all other methods.
`/v1/accounts/{account}/instance/{instance}/{version}/{name}`
don't want to have empty {version} in that path

File: pkg/apiserver/api/v1/instance.proto (hunk 10 above)
Comment (aborilov) [reply pkg/apiserver/api/v1/instance.proto:76]: actually, `Account` is also in `Instance.Metadata`, but there are no `version` and `instance` there

File: pkg/apiserver/api/v1/instance.proto (hunk 11 above)
Comment (nmiculinic) [reply pkg/apiserver/api/v1/instance.proto:52]: But this parameter is unnecessary. Maybe reverse the ordering as:
 
`/v1/accounts/{account}/instance/{instance}/{name}/{version}`

for all other API calls?

At the end of the day, I don't really care that much for REST API, since gRPC-web should be used by default.

File: pkg/apiserver/api/v1/instance.proto [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+  string version = 2;
+  // Name of the offering instance
+  string name = 3;
+  // Account indicate namespace of the project/account
+  string account = 4;
+}
+
+message InstanceListRequest {
+  // Offering name, i.e. couchdb.eu-west-1.team-a
+  string offering = 1;
```
Comment (nmiculinic) [inline pkg/apiserver/api/v1/instance.proto:61]: Can I list all instances within an account? Or is this something you'd like handling on the front-end?

File: pkg/apiserver/api/v1/instance.proto (hunk 7 above)
Comment (nmiculinic) [reply pkg/apiserver/api/v1/instance.proto:29]: Add offering information in the instance.

File: pkg/apiserver/internal/v1/instance.go [hunk 14]
```diff
@@ -0,0 +1,250 @@
...
+		client: c,
+		mapper: mapper,
+	}
+}
+func (o instanceServer) Create(ctx context.Context, req *v1.InstanceCreateRequest) (res *v1.Instance, err error) {
+	obj := &unstructured.Unstructured{}
+
+	gvk, err := o.gvkFromInstance(o.mapper, req.Offering, req.Version)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "creating instance: unable to get Kind: %s", err.Error())
```
Comment (nmiculinic) [inline pkg/apiserver/internal/v1/instance.go:57]: This isn't internal error, but argument error.

File: pkg/apiserver/api/v1/instance.proto (hunk 13 above)
Comment (aborilov) [reply pkg/apiserver/api/v1/instance.proto:61]: Do you mean all instances of all different offerings? In that case, no, current API can't do that, we can add this later, but it will be separate endpoint(`/v1/accounts/{account}/instances/`

File: pkg/apiserver/api/v1/instance.proto (hunk 11 above)
Comment (aborilov) [reply pkg/apiserver/api/v1/instance.proto:52]: but for the listing we have `"/v1/accounts/{account}/instances/{offering}/{version}"`, so we can't revers that

File: pkg/apiserver/api/v1/instance.proto (hunk 13 above)
Comment (nmiculinic) [reply pkg/apiserver/api/v1/instance.proto:61]: ok

File: pkg/apiserver/internal/v1/instance.go (hunk 14 above)
Comment (aborilov) [reply pkg/apiserver/internal/v1/instance.go]: done

**Comment by aborilov** [comment]
> Please add integration test for the services, under `/test/integration/apiserver.go`

done

File: pkg/apiserver/api/v1/instance.proto (hunk 7 above)
Comment (aborilov) [reply pkg/apiserver/api/v1/instance.proto]: done

File: pkg/apiserver/api/v1/instance.proto [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+
+import "google/api/annotations.proto";
+import "google/protobuf/empty.proto";
+
+import "meta.proto";
+
+message Instance {
+  ObjectMeta metadata = 1;
+  string offering = 2;
+  string spec = 3;
```
Comment (thetechnick) [inline pkg/apiserver/api/v1/instance.proto:29]: Could you please add a wrapper object for arbitrary data? (`status` and `spec` fields)
I would like to have the possibility to support multiple encoding styles the future.
e.g. `JSON` or `YAML` - maybe even nested protobuf

```
message RawObject {
  string encoding = 1;
  bytes data = 2;
}
```
Please use `JSON` as default for know, although I already see a use-case for the UI requesting the nested object as `YAML` (When editing the YAML directly)

File: pkg/apiserver/api/v1/instance.proto [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+option go_package = "v1";
+
+import "google/api/annotations.proto";
+import "google/protobuf/empty.proto";
+
+import "meta.proto";
+
+message Instance {
+  ObjectMeta metadata = 1;
+  string offering = 2;
```
Comment (thetechnick) [inline pkg/apiserver/api/v1/instance.proto:28]: Ok for this PR, but I see us changing this into a common `kind`/`apiVersion` field similar to the Kubernetes one.
Although we need to specify the mapping of Kubernetes apiVersion to "Our apiVersion", especially for these dynamic services.

File: pkg/apiserver/api/v1/instance.proto [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+  // Status - instance status
+  RawObject status = 4;
+}
+
+message InstanceList {
+  ListMeta metadata = 1;
+  repeated Instance items = 2;
+}
+
+message InstanceGetRequest {
```
Comment (jiachengxu) [inline pkg/apiserver/api/v1/instance.proto:42]: Since PR #426 is merged, can you please reuse/extend the Get/ListRequest in Service API?

Comment (aborilov) [reply pkg/apiserver/api/v1/instance.proto:42]: but Service API has additional params, that no present in `Get/ListRequest`

File: pkg/apiserver/api/v1/instance.proto (hunk 15 above)
Comment (aborilov) [reply pkg/apiserver/api/v1/instance.proto]: done

File: pkg/apiserver/api/v1/instance.proto (hunk 17 above)
Comment (jiachengxu) [reply pkg/apiserver/api/v1/instance.proto:42]: I think you can extend Get/ListRequest with new params. I also use Get/ListRequest in the Account PR.

Comment (aborilov) [reply pkg/apiserver/api/v1/instance.proto:42]: yes, because account service doesn't have any additional fields

Comment (jiachengxu) [reply pkg/apiserver/api/v1/instance.proto:42]: OK, so what is the reason to have a general List/GetRequest? I don't see any reason that we can not extend it to include params and use it everywhere

File: pkg/apiserver/api/v1/util.go [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+	"encoding/json"
+	"errors"
+	fmt "fmt"
+
+	"sigs.k8s.io/yaml"
+)
+
+const (
+	FormatJSON = "json"
+	FormatYAML = "yaml"
```
Comment (thetechnick) [inline pkg/apiserver/api/v1/util.go:29]:
```suggestion
	EncodingJSON = "json"
	EncodingYAML = "yaml"
```

Just a nit :)

Comment (aborilov) [reply pkg/apiserver/api/v1/util.go:29]: done

File: test/integration/servicecluster.go [hunk 19]
```diff
@@ -123,7 +123,7 @@ func newServiceClusterSuite(
 			},
 		}
 		require.NoError(t, managementClient.Create(ctx, catalogEntrySet))
-		require.NoError(t, testutil.WaitUntilReady(ctx, managementClient, catalogEntrySet, testutil.WithTimeout(2*time.Minute)))
+		require.NoError(t, testutil.WaitUntilReady(ctx, managementClient, catalogEntrySet, testutil.WithTimeout(time.Minute)))
```
Comment (jiachengxu) [inline test/integration/servicecluster.go:126]: Why do you change this timeout back?

File: pkg/testutil/framework.go [hunk 20]
```diff
@@ -232,6 +232,60 @@ func (f *Framework) NewFakeDB(name, namespace string) *fakev1alpha1.DB {
 	}
 }
 
+func (f *Framework) NewServiceCluster(name, namespace, secret string) *corev1alpha1.ServiceCluster {
```
Comment (jiachengxu) [inline pkg/testutil/framework.go:235]: How about we remove the Framework pointer receiver of these helper functions and add them in https://github.com/kubermatic/kubecarrier/blob/offering-watch-api/pkg/testutil/helpers.go, the reason is because:
1. the framework receiver is unused in these helper functions, and doesn't make sense to have it.
2. Then you don't need to change the of the following offeringService, regionService, providerService in the following just for using helper functions.
3. Our unit tests can also use them in the future.

File: test/integration/apiserver.go [hunk 21]
```diff
@@ -197,22 +197,170 @@ func newAPIServer(f *testutil.Framework) func(t *testing.T) {
...
 	}
 }
 
+func instanceService(ctx context.Context, conn *grpc.ClientConn, managementClient *testutil.RecordingClient, f *testutil.Framework) func(t *testing.T) {
+	return func(t *testing.T) {
+		serviceClient, err := f.ServiceClient(t)
+		require.NoError(t, err, "creating service client")
+		ctx, cancel := context.WithCancel(context.Background())
+		t.Cleanup(cancel)
+		managementClient, err := f.ManagementClient(t)
```
Comment (jiachengxu) [inline test/integration/apiserver.go:223]: I don't think you need the managementClient and create a new context in the function, you can just use them by passing from the function, right?

File: test/integration/servicecluster.go (hunk 19 above)
Comment (aborilov) [reply test/integration/servicecluster.go:126]: because I thought that the problem was with `catalogentryset` cleanup, but apparently it also needs to be increased here

File: test/integration/apiserver.go (hunk 21 above)
Comment (aborilov) [reply test/integration/apiserver.go:223]: you are right

File: pkg/testutil/framework.go (hunk 20 above)
Comment (aborilov) [reply pkg/testutil/framework.go:235]: then I will move all `NewSomething` helpers

Comment (jiachengxu) [reply pkg/testutil/framework.go:235]: Yeah, would be really great, I have moved the NewProviderAccount and stuff in my account PR also.

Comment (jiachengxu) [reply pkg/testutil/framework.go:235]: You can move helpers that are used by your PR, and then we move the rest of helpers gradually :)

Comment (aborilov) [reply pkg/testutil/framework.go:235]: too late, already moved everything :)

Comment (aborilov) [reply pkg/testutil/framework.go:235]: but I still need `func instanceService(ctx context.Context, conn *grpc.ClientConn, managementClient *testutil.RecordingClient, f *testutil.Framework) func(t *testing.T) {
` in my tests, so have to change the signatures of the following offeringService, regionService, providerService

File: test/integration/apiserver.go (hunk 21 above)
Comment (aborilov) [reply test/integration/apiserver.go:221]: done

File: pkg/testutil/framework.go (hunk 20 above)
Comment (aborilov) [reply pkg/testutil/framework.go]: done
